### DISCUSSION TITLE
why Provider always at the top level？

### DISCUSSION THREAD

**Discussion by wangcongyi** [TanStack/query#4394] [Q&A]
I have an increasingly complex application.
One day I need redux, and add the Provider component at the top level,
On another day I need react-query, and add the QueryClientProvider component at the top level too.
etc...
After many times the code is like this:

```js
<ConfigProvider locale={zhCN}>
  <QueryClientProvider client={queryClient}>
    <Provider store={store}>
      <Router basename={basename}>
        <Layout />
      </Router>
    </Provider>
  </QueryClientProvider>
</ConfigProvider>
```

I worry that it will increase more and more in the future.
so Why provider (including other tools or libraries ) always at the top level？
Is this for convenience or what? Why do we need provider?

**Comment by TkDodo** [comment]
> I worry that it will increase more and more in the future.

Why? This is how react works, it's not a problem.

Providers allow to inject dependencies. That's how `useSelector` from redux gets access to the store you have created, and how `useQuery` also gets access to the cache that you have created when a QueryClient is created.

**Reply by wangcongyi** [reply]
yes, but the provider is kind of nested way, I don't know why the popular libraries of react all choose Providers.
Is the provider the only way? Is there a better way to write it instead of the nested form?

**Reply by TkDodo** [reply]
it's _the way_ to do dependency injection in react ...

**Reply by TkDodo** [reply]
maybe you need to hear from Dan directly that this is fine...?

https://twitter.com/dan_abramov/status/1563307506482696192

**Reply by wangcongyi** [reply]
ok, thank you so much
